### PR TITLE
[ext] Add typed metadata support to dagster-ext

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Adds a `metadata_type` field to `report_asset_metadata`. This takes a string literal (specified in the `ExtMetadataType` type alias). These strings correspond to static methods on `MetadataValue` that construct metadata values of the target type.

## How I Tested These Changes

Modified unit tests.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16383** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16383" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16383?utm_source=stack-comment).

File: python_modules/dagster-ext/dagster_ext/__init__.py [hunk 1]
```diff
@@ -668,15 +698,23 @@ def extras(self) -> Mapping[str, Any]:
     # ##### WRITE
 
     def report_asset_metadata(
-        self, label: str, value: Any, asset_key: Optional[str] = None
+        self,
+        label: str,
+        value: Any,
+        type: Optional[ExtMetadataType] = None,  # noqa: A002
```
Comment (schrockn) [inline python_modules/dagster-ext/dagster_ext/__init__.py:704]: `ruff` might have a point here. name it `metadata_type`?

File: python_modules/dagster-ext/dagster_ext/__init__.py [hunk 2]
```diff
@@ -668,15 +698,23 @@ def extras(self) -> Mapping[str, Any]:
...
+        asset_key: Optional[str] = None,
     ) -> None:
         asset_key = _resolve_optionally_passed_asset_key(
             self._data, asset_key, "report_asset_metadata"
         )
         label = _assert_param_type(label, str, "report_asset_metadata", "label")
         value = _assert_param_json_serializable(value, "report_asset_metadata", "value")
+        _type = _assert_opt_param_value(
```
Comment (schrockn) [inline python_modules/dagster-ext/dagster_ext/__init__.py:712]: yeah let's rename this

File: python_modules/dagster/dagster/_check/__init__.py [hunk 3]
```diff
@@ -882,6 +882,31 @@ def is_list(
...
 
+# ########################
+# ##### LITERAL
+# ########################
+
+
+def literal_param(
+    obj: T, param_name: str, values: Sequence[object], additional_message: Optional[str] = None
+) -> T:
+    additional_message = " " + additional_message if additional_message else ""
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:893]: nit: could move this inside of conditional or delete and rely on `_param_value_mismatch_exception`  [1]

File: python_modules/dagster/dagster/_check/__init__.py [hunk 4]
```diff
@@ -882,6 +882,31 @@ def is_list(
...
+    if obj not in values:
+        raise _param_value_mismatch_exception(obj, values, param_name, additional_message)
+    return obj
+
+
+def opt_literal_param(
+    obj: T, param_name: str, values: Sequence[object], additional_message: Optional[str] = None
+) -> T:
+    additional_message = " " + additional_message if additional_message else ""
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:902]: [1]

File: python_modules/dagster/dagster/_core/ext/context.py [hunk 5]
```diff
@@ -38,12 +40,17 @@ def handle_message(self, message: ExtMessage) -> None:
         elif message["method"] == "log":
             self._handle_log(**message["params"])  # type: ignore
 
-    def _handle_report_asset_metadata(self, asset_key: str, label: str, value: Any) -> None:
+    def _handle_report_asset_metadata(
+        self, asset_key: str, label: str, value: Any, type: ExtMetadataType  # noqa: A002
+    ) -> None:
         check.str_param(asset_key, "asset_key")
         check.str_param(label, "label")
+        check.opt_literal_param(type, "type", get_args(ExtMetadataType))
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/ext/context.py:48]: does `get_args` cache?

File: python_modules/dagster-ext/dagster_ext/__init__.py [hunk 6]
```diff
@@ -94,6 +96,23 @@ class ExtDataProvenance(TypedDict):
...
+    "json",
+    "md",
+    "float",
+    "int",
+    "bool",
+    "dagster_run",
+    "asset",
+    "table",
+    "null",
+]
```
Comment (schrockn) [inline python_modules/dagster-ext/dagster_ext/__init__.py:113]: where do we guarantee that these all line up to valid string literals in dagster core?

File: python_modules/dagster/dagster/_core/ext/context.py [hunk 7]
```diff
@@ -38,12 +40,17 @@ def handle_message(self, message: ExtMessage) -> None:
...
         key = AssetKey.from_user_string(asset_key)
         output_name = self._context.output_for_asset_key(key)
-        self._context.add_output_metadata({label: value}, output_name)
+        # metadata type values correspond to static methods on MetadataValue
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/ext/context.py:51]: we should not be relying on keeping static methods on `MetadataValue` up-to-date and the members of ExtMetadataType synced via convention alone. Needs to be guaranteed by some sort of code artifact or a unit test that ensures that they correspond by using reflection apis

**Review by schrockn — CHANGES_REQUESTED** [review]
Looks good.

Convention around static methods on `MetadataValue` is too sketchy, especially in light of extremely thin test suite. 

All the possible literal types should be tested, and some sort of construct needs to exist to keep them in sync with the corresponding translation in the orch process.

File: python_modules/dagster/dagster/_core/ext/context.py (hunk 5 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/ext/context.py:48]: I don't know but it's not slow, compared to directly looking up a list on an object (10000 times each):

```
Time taken for get_args_from_literal: 0.002764 seconds
Time taken for get_array_from_object: 0.001988 seconds
```

File: python_modules/dagster-ext/dagster_ext/__init__.py (hunk 6 above)
Comment (smackesey) [reply python_modules/dagster-ext/dagster_ext/__init__.py]: they are now tested

File: python_modules/dagster/dagster/_check/__init__.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_check/__init__.py:893]: oh yeah whoops that was left over from before I factored out `_param_value_mismatch_exception`

File: python_modules/libraries/dagster-aws/dagster_aws/ext.py [hunk 8]
```diff
@@ -12,7 +12,7 @@
 
 
 class ExtS3MessageReader(ExtBlobStoreMessageReader):
-    def __init__(self, *, interval: int = 10, bucket: str, client: boto3.client):
+    def __init__(self, *, interval: float = 10, bucket: str, client: boto3.client):
```
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/ext.py:15]: unrelated?

File: python_modules/dagster/dagster/_check/__init__.py [hunk 9]
```diff
@@ -1645,6 +1670,21 @@ def _element_check_error(
...
 
+def _param_value_mismatch_exception(
+    obj: object,
+    values: Sequence[object],
+    param_name: str,
+    additional_message: Optional[str] = None,
+    allow_none: bool = False,
+) -> ParameterCheckError:
+    allow_none_clause = " or None" if allow_none else ""
+    additional_message = " " + additional_message if additional_message else ""
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:1681]: not a huge fan of this sort of boolean param. Easy to screw up and non-intuitive

File: python_modules/dagster-ext/dagster_ext/__init__.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster-ext/dagster_ext/__init__.py:703]: would like this changed before commit

File: python_modules/dagster/dagster/_check/__init__.py (hunk 9 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_check/__init__.py]: It follows an existing pattern in `check` (see `_param_class_mismatch_exception` and IMO is intuitive in the limited context in which this method is used.

**Comment by smackesey** [comment]
In the process of updating this PR to address feedback I discovered that multiple calls to `report_asset_metadata` for the same asset did not work. This turns out to be because `add_output_metadata` (the `OpExecutionContext` counterpart of `report_asset_metadata`) had a restriction that it could only be called once-per-output. I'm not sure of the justification or origin for this restriction, but I tried removing it in this PR (changing to have repeated calls merge the additional passed metadata) and it didn't break anything except a test explicitly checking for an error.

Barring some good reason I'm unaware of for the restriction, this seems preferable to me to requiring users to only invoke `report_asset_metadata` once per asset or implementing some kind of metadata buffering, which are the alternatives that come to mind.

**Comment by schrockn** [comment]
@smackesey a more conservative I recommend is calling a new internal-only method called `accumulate_output_metadata` or something and leaving the semantics of `add_output_metadata` in place. Then we can change the behavior of `add_output_metadata` in a separate PR that can be considered on its own, reverted separately, and so forth.

**Comment by smackesey** [comment]
>a more conservative approach I recommend is calling a new internal-only method called accumulate_output_metadata or something and leaving the semantics of add_output_metadata in place. Then we can change the behavior of add_output_metadata in a separate PR that can be considered on its own, reverted separately, and so forth.

done and here is the separate PR: https://github.com/dagster-io/dagster/pull/16391

File: python_modules/dagster/dagster/_core/execution/context/compute.py [hunk 10]
```diff
@@ -410,6 +410,20 @@ def add_metadata_two_outputs(context) -> Tuple[str, int]:
...
+        mapping_key: Optional[str] = None,
+    ) -> None:
+        metadata = check.mapping_param(metadata, "metadata", key_type=str)
+        output_name = check.opt_str_param(output_name, "output_name")
+        mapping_key = check.opt_str_param(mapping_key, "mapping_key")
+
+        self._step_execution_context.add_output_metadata(
+            metadata=metadata, output_name=output_name, mapping_key=mapping_key, merge=True
+        )
+
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/context/compute.py:426]: can you add an `_add_output_metadata` helper method that takes the bool and leave the public one unmodified?

**Review by schrockn — CHANGES_REQUESTED** [review]
Let's leave the public method signature for `add_output_metadata` unchanged

File: python_modules/dagster/dagster/_core/execution/context/compute.py (hunk 10 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/context/compute.py:426]: `StepExecutionContext.add_output_metadata` isn't public, it's accessed through the public `OpExecutionContext.add_output_metadata`, which doesn't have any modifications in this PR.

Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/context/compute.py:426]: Or do you mean public just in the Python (not user-facing) sense? `StepExecutionContext.add_output_metadata` while python-public is an internal API so it seems find to add a switch?

**Comment by schrockn** [comment]
Ah I misread that! Great point

**Review by schrockn — APPROVED** [review]
